### ISSUE TITLE
Feature Request: Tutorial on using 1am

### ISSUE THREAD

**Issue by jgarte** [LispCookbook/cl-cookbook#476]
[1am](https://github.com/lmj/1am) is a classic CL testing framework that fits on a page.

`1am` is also a good testing framework for beginners because it is easy to learn for small projects.

I'd like to see a tutorial using it.

Here are some draft ideas for content:

```lisp
(defun my-length (l)
  "Takes the length of a list or vector."
  (etypecase l
    (list (list-length l))
    (vector (array-dimension l 0))))

(defun test-my-length ()
  "Testing my-lenth with 1am."
  (1am:is (= 12 (my-length (alexandria:iota 12))))
  (1am:is (= 12 (my-length (coerce (alexandria:iota 12) 'vector))))

  ;; Check that passing a keyword signals an error.
  (1am:signals type-error (= :hahaha (my-length (coerce (alexandria:iota 12) 'vector)))) 
  (1am:signals type-error (= :hahaha (my-length (alexandria:iota 12)))))

(test-my-length)
```

**Comment by jgarte** [comment]
phoe had mentioned on [matrix](https://matrix.to/#/#common-lisp:matrix.org) how he decides on a test framework:

> my two eurocents in the topic; how dumb is the code I am testing?

> the code is dumber than me - DEFUN over ASSERT
the code is dumber than average and also tiny - 1am
the code is of average or higher non-dumbness, or there's more than one page of it - parachute

**Comment by vindarel** [comment]
mmmh but we have a good tutorial for a good and popular testing framework (5am), so not sure a review of 1am would fit in the Cookbook.

I'll add a "wontfix" tag, but happy to see more discussion.

**Comment by jgarte** [comment]
Hi @vindarel 

As I mentioned in the example above, phoe uses a particular testing framework based on "how dumb is the code I am testing?": 

> my two eurocents in the topic; how dumb is the code I am testing?

> the code is dumber than me - DEFUN over ASSERT
> the code is dumber than average and also tiny - 1am
> the code is of average or higher non-dumbness, or there's more than one page of it - parachute

5am probably fits into the category of "the code is of average or higher non-dumbness ...".

I agree with phoe here in using the right testing tool for the job in contrast to the Python community, where most people just use [pytest](https://github.com/pytest-dev/pytest), by de facto. 

In other words, I'm not sure if 5am is the "correct hammer" for every job and in the spirit of quick recipes, "DEFUN over ASSERT" might be the most elegant solution, especially for a beginner where 5am would be overkill for their purposes.

> mmmh but we have a good tutorial for a good and popular testing framework (5am), so not sure a review of 1am would fit in the Cookbook.

Is this because you'd like the cookbook to show just one way of doing things? 

If so, do you think that we should state a project goal in the README that we should not duplicate concepts to show different approaches taken as it is out of scope for the project?

**Comment by vindarel** [comment]
Good point in that talking about 5AM up front on the testing page is quite much. Now I think it'd great if we introduced simpler testing patterns -the opportunity to teach some CL.

Random thoughts on 1am: you can write a tutorial outside the Coobook to educate us all ;) It has no Github activity and even if it's done© this gives a very bad image of CL to newcomers; 5AM has some plugins; 1AM doesn't mention fixtures, something I judge basic for a test framework; Phoe is an experienced lisper and experienced lispers can form other opinions after they have learned the basics and the popular tools!

**Comment by vindarel** [comment]
> What simpler testing patterns were you thinking of more or less?

First, `assert` as you said, also the `#+my-test` trick, using closures, unwind-protect, recalling the interactive debugger features, maybe a simple macro.
